### ISSUE TITLE
Test

### ISSUE THREAD

**Issue by Kulikovpavel** [Panek1207/QuickFacts#1]
import numpy as np
a = np.random.randint(-1, 2, (200, 200))
U, s, V = np.linalg.svd(a, full_matrices=False)
print(U.shape, V.shape, s.shape)

S = np.diag(s)
np.allclose(a, np.dot(U, np.dot(S, V)))
print(a)

**Comment by Ilya1510** [comment]
import numpy as np


def algorithm_svd(a):
    mat_u, s, mat_v = np.linalg.svd(a, full_matrices=False)
    mat_s = np.diag(s)
    d = np.linalg.matrix_rank(a)
    mat_u = mat_u[0: mat_u.shape[0], 0: d]
    mat_s = mat_s[0: d, 0: d]
    mat_v = mat_v[0: d, 0:mat_v.shape[1]]
    if not np.allclose(a, np.dot(mat_u, np.dot(mat_s, mat_v))):
        print("Bugaga")
    return [np.dot(mat_u, mat_s), mat_v]


def give_advise(data, mat_u, mat_f, user_id):
    marks = np.dot(mat_u[user_id], mat_f)
    for i in range(mat_f.shape[1]):
        sum = 0
        for j in range(mat_f.shape[0]):
            sum += mat_u[j][0] * mat_f[j][i]
        print(sum)
    val = 0.0
    film_id = 0
    print(mat_u[user_id])
    print(data)
    print(mat_u)
    print(mat_f)
    for i in range(marks.shape[0]):
        if data[user_id][i] == 0 and val < marks[i]:
            val = marks[i]
            film_id = i
    print(film_id)
    print(marks)

m = [[1, -1, 0, 0, 1, 1],
     [1, -1, 1, -1, 1, 1],
     [1, -1, 1, -1, 1, 1],
     [1, -1, 1, -1, 1, 1],
     [-1, 1, 1, -1, 1, 1],
     [-1, 1, 1, -1, 1, 1],
     [-1, 1, 1, -1, 1, 1],
     [1, 1, 1, -1, 1, 1],
     [1, 1, 1, -1, 1, 1],
     [1, 1, 1, -1, 1, 1]
     ]
m1, m2 = algorithm_svd(m)
give_advise(m, m1, m2, 0)

**Comment by Kulikovpavel** [comment]
Да, завтра нужно будет реализовать инкрементный алгоритм, машинного обучения параметров матриц разложения, и вот оно сработает.

**Comment by Kulikovpavel** [comment]
http://stackoverflow.com/questions/8847548/howto-create-recommendations-with-a-incremental-svd-recommender-system

**Comment by Kulikovpavel** [comment]
https://classes.soe.ucsc.edu/cmps242/Winter08/proj/serdar_talk.pdf

**Comment by Ilya1510** [comment]
`import numpy as np
import copy


def algorithm_svd(a):
    mat_u, s, mat_v = np.linalg.svd(a, full_matrices=False)
    mat_s = np.diag(s)
    d = np.linalg.matrix_rank(a)
    mat_u = mat_u[0: mat_u.shape[0], 0: d]
    mat_s = mat_s[0: d, 0: d]
    mat_v = mat_v[0: d, 0:mat_v.shape[1]]
    if not np.allclose(a, np.dot(mat_u, np.dot(mat_s, mat_v))):
        print("Bugaga")
    return [np.dot(mat_u, mat_s), mat_v]


def dist(mat1, mat2):
    diff = 0.0
    for i in range(min(mat1.shape[0], mat2.shape[0])):
        for j in range(min(mat1.shape[1], mat2.shape[1])):
            diff += abs(mat1[i][j] - mat2[i][j])
    return diff


def sum_square(mat):
    sum = 0.0
    for i in range(mat.shape[0]):
        for j in range(mat.shape[1]):
            sum += mat[i][j]
    return sum


def study(data, mat_u, mat_f):
    learning_rate = 0.03
    coef_regular = 0.001
    while True:
        mat_u_new = copy.deepcopy(mat_u)
        mat_f_new = copy.deepcopy(mat_f)
        for i in range(mat_u.shape[0]):
            for j in range(mat_f.shape[1]):
                if data[i][j] == 0:
                    continue
                delta = 0.0
                for k in range(mat_u.shape[1]):
                    delta -= mat_u[i][k] * mat_f[k][j]
                delta += data[i][j]
                delta = delta
                #delta += coef_regular * (sum_square(mat_u) + sum_square(mat_f) - sum_square(mat_u_new) - sum_square(mat_f_new))
                for k in range(mat_u.shape[1]):
                    mat_u_new[i][k] = mat_u[i][k] + learning_rate * delta * mat_f[k][j]
                    mat_f_new[k][j] = mat_f[k][j] + learning_rate * delta * mat_u[i][k]
        #print(mat_u)
        #print(mat_f)
        #print(np.dot(mat_u, mat_f))
        print(dist(data, np.dot(mat_u, mat_f)))
        if dist(np.dot(mat_u, mat_f), np.dot(mat_u_new, mat_f_new)) < 0.0001:
            break
        mat_u = mat_u_new
        mat_f = mat_f_new
    return mat_u, mat_f


def give_advise(data, mat_u, mat_f, user_id):
    marks = []
    for i in range(mat_f.shape[1]):
        sum = 0
        for j in range(mat_f.shape[0]):
            sum += mat_u[user_id][j] * mat_f[j][i]
        marks.append(sum)
    val = 0.0
    film_id = 0
    for i in range(len(marks)):
        if data[user_id][i] == 0 and val < marks[i]:
            val = marks[i]
            film_id = i
    print(film_id)
    print(marks)

m = np.array([[0, 0, 1, 1, 1, 1],
     [0, 0, 1, 0, 1, 0],
     [1, -1, 1, 1, 1, -1],
     [1, -1, 0, 1, 1, -1],
     [1, -1, 1, 1, 1, -1],
     [1, -1, 1, 0, 1, -1],
     [1, -1, 1, 1, 1, -1],
     [1, 0, 1, 1, 1, -1],
     [1, -1, 0, 1, 1, -1],
     [1, -1, 1, 1, 1, 1],
     ])
m1, m2 = algorithm_svd(m)
m1 = np.full(m1.shape, 0.1)
m2 = np.full(m2.shape, 0.1)
m1, m2 = study(m, m1, m2)
give_advise(m, m1, m2, 0)
`
